### ISSUE TITLE
Proposal for Reference Gallery

### ISSUE THREAD

**Issue by philippjfr** [holoviz/panel#4305]
Our reference gallery serves as a valuable visual index of all the different components that Panel has to offer. Its strength lies is the fact that it is both browsable and searchable, unlike a more classical API reference. However once we reach the actual individual reference guides it becomes a little messier. They all follow a general structure that roughly follows:

- Description of what the component is and what it does.
- Some high-level comments about usage
- A list of parameters that are specific to that component and what they do
- Now come a number of usage example that highlight basic usage and then a non-exhaustive list of examples
- Lastly most Pane and Widget components have a `Controls` section that uses a JS linked form to allow you to play with at least some of the parameters.

There are two main problem with this approach:

1. The high-level description and list of parameters and their descriptions are completely disconnected from the docstrings and the code so it's easy for them to get out of sync and they are annoying to update.
2. Usage examples aren't standalone and if you want to contribute one you have to make it fit into the flow of the overall page.

## Proposal

What I propose is that we automatically generate these pages by pulling content from a number of places:

1. The description of the component should be pulled directly from the docstring
2. The list of parameters and their descriptions should be automatically generated from the code
3. We automatically embed a collapsible section of any inherited parameters
4. The examples section will be generated automatically by pulling in small, self-contained examples.

For 4. I envision the individual examples as very small how-tos that demonstrate one thing about that particular component. We would then need some mechanism to tell it which of these examples would get rendered directly on the page and which would live in a small sub-gallery.

## Sketch

<img width="409" alt="Screen Shot 2023-01-20 at 11 38 58" src="https://user-images.githubusercontent.com/1550771/213675702-36c8ce1b-e472-4fb3-b348-3bda56efc6eb.png">

**Comment by maximlt** [comment]
I think 1., 2. and 3. would be nice improvements compared to the current state.

If you add to that the list of methods supported by the component (I guess in a collapsed container), doesn't this mean that there's actually no need of the component to be in the API reference? (Making it easier to merge the API reference and visual reference)

I am much more doubtful about the usefulness of 4:

- I don't think the Reference Gallery in its current state is so bad (e.g. compared to how the user guides, the API reference and to some extent the Gallery which is a mix of How-Tos and Demos) that it needs such a big change. The Reference Gallery is currently and imo the best section of the whole documentation.
- I really like to have every thing in a single page and do CTRL+F to look for what I'm after. I do it very often. Being able to quickly scroll over a component's page to see what it offers is also nice.
- This is additional code to develop and maintain (in NBSite I guess) and this is extra layer of complexity in building the docs (this will get even more complex if this is meant to be adapted to HoloViews and hvPlot)
- That ship has already partially sailed for Panel, at least temporarily, but the ability for users to download a page of the website as a notebook and run it locally isn't going to be made easier if the examples for a component are scattered across many small pages.

> Usage examples aren't standalone

The examples not being standalone can be dealt with by following more strict guidelines about how to write them in the page. This is exactly what [Plotly recommends](https://github.com/plotly/plotly.py/tree/doc-prod/doc) in its guidelines for contributing to their docs:

```
We try to write short, standalone and (almost) self-explaining examples. Most examples should focus on a single feature.

Checklist

    Each example should have a clear title (titles are used for the navigation bar and indexed by search engines)
    Package imports should be called in the same cell as the example, so that it is possible to copy-paste a single cell to reproduce the example.
    Variable names should be consistent with other examples, for example use fig for a Figure object, df for a pandas dataframe, etc.
    Examples should not be too long to execute (typically < 10s), since the doc is built as part of the continuous integration (CI) process. Examples taking longer to execute should be discussed in a new issue to decide whether they can be accepted.
```

> if you want to contribute one you have to make it fit into the flow of the overall page.

I would argue that whatever example you write for a component has to fit into some flow. Having them all in a single page actually helps grouping them in sections that make sense, and ordering them given their usefulness.



In short, as a user I think this change wouldn't bring me much benefit, if any at all, and as a maintainer I think that it would be a lot of work.

---

In the (very nice !) sketch you added a *Discover more* section. Instead of containing a gallery of small examples, it could link to all the places in the documentation where the component is used.

**Comment by MarcSkovMadsen** [comment]
One comment is that the reference examples in its current state in my experience is the most used place to go. So we should not throw it away before we clearly have something better.

For me the main pains are 

- the intro section can get out of sync with the components docstring, parameters and functionality
- the examples are not self contained
- the reference guide is not live and opening it in a live notebook takes time.

I have also hoped that

- there would be a nice `pn.help(component)` that we could use at the top of the notebook instead of writing things manually.
- the examples would be self contained with in a small section.
- it could be very, very easy and quick (1-2 secs) to open the notebook on Panelite or Binder.

**Comment by MarcSkovMadsen** [comment]
An alternative would be separating the reference documentation.

You could in one place provide great auto generated documentation. Then you could in another place have an `example gallery` of very small examples. The gallery would be highly searchable. And the examples open quickly on Panelite.

**Comment by maximlt** [comment]
> Then you could in another place have an example gallery of very small examples.

Would that gallery contain all the examples of all the components? Or that would but just one gallery per component, as suggested in the OP?

**Comment by droumis** [comment]
I agree with @philippjfr. 

And I do think the Reference Gallery should be merged with the Reference API.


I also think Maxime's observation is a fair point:

> I really like to have every thing in a single page and do CTRL+F to look for what I'm after. I do it very often. Being able to quickly scroll over a component's page to see what it offers is also nice.

Maybe the example cards, as depicted in OP, could include a short description that could be picked up by CTRL+F-ing on the page.

**Comment by MarcSkovMadsen** [comment]
Totally agree on `Automate parameter and method documentation: I'd argue for doing so at the Param level -- let's make a nice Parameterized HTML repr and use that everywhere! Way overdue.`.

This will also be useful for any user of Param.

**Comment by MarcSkovMadsen** [comment]
Regarding format. It would be best if the code examples we write can be tested. Its not the case currently for the `.md` how to guides we write and that will hit us at some point in time.

**Comment by philippjfr** [comment]
@MarcSkovMadsen that is not true. The code in the pyodide directives of the markdown files is executed as part of the unit tests.

**Comment by maximlt** [comment]
> Make visual examples be self-contained and independent of each other. I think we agree on the principle, but if we further specify how that works I think everyone could be happy:
> 
>     * Store self-contained examples as independent snippets of Markdown or small non-narrative Jupyter notebooks
> 
>     * When building the website, inline some curated subset of these examples directly in the page for high visibility and scroll searching
> 
>     * Embed the other examples collapsed into the page in a way that supports CTRL-F, either on their descriptions or on their collapsed contents (not sure exactly how that works with browsers)
> 
>     * When building the site, generate some downloadable Jupyter notebook containing all this info for people to explore offline or in JupyterLite, as in e.g. https://scikit-image.org/docs/stable/auto_examples/numpy_operations/plot_structuring_elements.html . Maintainers wouldn't edit such a notebook, only the independent examples, but users could play with it easily.

It seems to be that would be a lot of work I would like to better understand what we are trying to fix or improve:
- did users complain about the examples not being standalone?
- what other complaints did we get besides what Marc mentioned above? (the examples not being live, which is an orthogonal problem compared to what is being discussed here)
- Where is there actually how-to/learning material in the current Reference Gallery?
- what do the maintainers complain about?

> I think if we follow this approach we ensure that the reference gallery contains no user guide or howto information, only the actual reference material along with atomic examples, while still being searchable as the pages currently are.

It seems to me the peer-review process can ensure that, more than the system we choose.

I would like us to avoid cluttering the How-To with things that would also make sense on a reference page, if that info is specific to that component only. E.g. "How do I highlight Tabulator cells" should IMO be in the reference page of Tabulator, and not have its own section if the How-To.

**Comment by jbednar** [comment]
I agree that reference material about an object should be in the reference page for that object; I don't think anyone would disagree with that. The goal is that most such documentation should be in the docstrings of the object, with highly visual stuff appearing in the examples for this component, showing an example input and an example output. 

The reason for examples to be standalone is so that it makes sense to curate them outside of the structure of a reference gallery page. I.e., the point is to take our current reference gallery content, and to maintain it in the right location, not to eliminate it. E.g. the parameter docs shouldn't be on a narrative page, they should be automatically pulled from the parameter docstrings. The component docs shouldn't be on a narrative page, they should be automatically pulled from the component docstrings. And then what's left shouldn't be a narrative, but a series of examples, so that we don't end up with a shadow duplicate structure of docs , mirroring the API reference while duplicating aspects of it. That way instead of having API docs that are authoritative but lack examples, and reference gallery pages that have good examples but are not authoritative, there can be one primary reference for each component, both authoritative and giving access to all examples for that component. I think the way to achieve that is to define examples as purely self-contained items that make sense to link to from anywhere. To do that we have to be able to introduce them consistently as "here are some examples" as opposed to (as now) fitting everything into a single narrative. In practice I don't see Panel's reference gallery losing its current contents; instead I see it becoming authoritative about all aspects of the component, by embedding the examples it already has in the context of full and complete auto-generated information about that component.

**Comment by maximlt** [comment]
Thanks for the detailed write up. To be clear, I'm in favor of 1., 2., and 3. suggested by Philipp; I am doubtful about 4.

> To do that we have to be able to introduce them consistently as "here are some examples" as opposed to (as now) fitting everything into a single narrative.

I still would like to hear what explicit problems we're trying to solve with 4 :) ? What did our users complain about? Is it an improvement to offer our users a bunch of randomly organized examples instead of examples organized in a single narrative?

If the problem we are trying to fix is that examples are not standalone, I think this is pretty low priority as, as someone reading some docs I do not expect every single code block to be standalone. That's nice to have for sure, but I'm also fine if at the beginning of the page there's a comment that tells me "to reproduce each example you will need to include this setup code block", even more so if this pattern is consistently applied across the docs. This is something we could do right now without to invent a new system. Or we could decide to make them all standalone and this is also something we could do right now, it's just a bunch of copy/paste (that would also need to be done if we decide to go with the automatic examples system, to create each one of these examples).

I would also like to know which files will actually need to exist and where they are going to be stored on the repo? For instance, is the reference page of a component (e.g. `Select`) going to be entirely automatically generated at the build step? If it exists before the build, what is it going to contain?

And to be clear on what I suggest, here's what I think the reference page of a component could look like. That page would exist and be in the doc/ folder. I assume that Philipp intents the examples to be executable with Pyodide so this is going to be a markdown file.

```

# WidgetName

:::{directive that automatically and nicely displays the widget description, parameters, inherited parameters and methods}
    pn.widgets.widgetname
:::

## Examples

### Example 1

Some text

:::pyodide
some code
:::

### Example 2
...

```

Having the examples manually in-lined in the reference as it is right now:
- allows us to group and order the examples in something that makes sense (for sure some examples are more important that others)
- allows us to add some bits of context when that makes sense. E.g. the Tabulator widget provides multiple ways to filter a table (python callbacks, client-side), a simple comment, put before a group of examples on filtering, that says that there are multiple approaches and compares them, makes sense to me. That is imo reference-content material and I don't know where else this would be best documented.
- does not require to setup a new system
